### PR TITLE
Fix 'Provide the capability to access browser information from test code' (closes #481)

### LINKED ISSUE #481: Provide the capability to access browser information from test code
Make browser information available in the test controller - add the `t.browserInfo` property.

There are various use cases for this feature, including conditional test actions and assertions:

```js
test('My Test', async t => {
    if(t.browserInfo.name === 'IE')
        await t.click('#some-div');
});
```

### PR DESCRIPTION
**Pull request by Farfurix**
https://github.com/DevExpress/testcafe/issues/481

### Changes
1. Add `t.browser`.

### `t.browser` examples

<details>

<summary>testcafe "saucelabs:Internet Explorer@11.0:Windows 7" test.js (Internet Explorer 11.0 / Windows 7)</summary>

```
{
  alias: 'saucelabs:Internet Explorer@11.0:Windows 7',
  headless: false,
  name: 'Internet Explorer',
  version: '11.0',
  platform: 'desktop',
  os: { name: 'Windows', version: '7' },
  engine: { name: 'Trident', version: '7.0' },
  prettyUserAgent: 'Internet Explorer 11.0 / Windows 7',
  userAgent: 'Mozilla/5.0 (Windows NT 6.1; WOW64; Trident/7.0; rv:11.0) like Gecko'
}

```

</details>

<details>

<summary>testcafe chrome test.js (Chrome 77.0.3865.120 / Windows 10)</summary>

```
{
  alias: 'chrome',
  headless: false,
  name: 'Chrome',
  version: '77.0.3865.120',
  platform: 'desktop',
  os: { name: 'Windows', version: '10' },
  engine: { name: 'Blink' },
  prettyUserAgent: 'Chrome 77.0.3865.120 / Windows 10',
  userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.120 Safari/537.36'
}
```

</details>

<details>

<summary>testcafe "chrome:headless --no-sandbox" test.js (Chrome 77.0.3865.120 / Windows 10)</summary>

```
{
  alias: 'chrome:headless --no-sandbox',
  headless: true,
  name: 'Chrome',
  version: '77.0.3865.120',
  platform: 'desktop',
  os: { name: 'Windows', version: '10' },
  engine: { name: 'Blink' },
  prettyUserAgent: 'Chrome 77.0.3865.120 / Windows 10',
  userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) HeadlessChrome/77.0.3865.120 Safari/537.36'
}
```

</details>

<details>

<summary>testcafe firefox:headless:disableMultiprocessing=true test.js (Firefox 69.0 / Windows 10)</summary>

```
{
  alias: 'firefox:headless:disableMultiprocessing=true',
  headless: true,
  name: 'Firefox',
  version: '69.0',
  platform: 'desktop',
  os: { name: 'Windows', version: '10' },
  engine: { name: 'Gecko', version: '20100101' },
  prettyUserAgent: 'Firefox 69.0 / Windows 10',
  userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:69.0) Gecko/20100101 Firefox/69.0'
}
```

</details>

<details>

<summary>testcafe "saucelabs:iPad Air Simulator@12.2" test.js (Safari 12.1 / iOS 12.2)</summary>

```
{
  alias: 'saucelabs:iPad Air Simulator@12.2',
  headless: false,
  name: 'Safari',
  version: '12.1',
  platform: 'tablet',
  os: { name: 'iOS', version: '12.2' },
  engine: { name: 'WebKit', version: '605.1.15' },
  prettyUserAgent: 'Safari 12.1 / iOS 12.2',
  userAgent: 'Mozilla/5.0 (iPad; CPU OS 12_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.1 Mobile/15E148 Safari/604.1'
}
```

</details>

<details>

<summary>testcafe "saucelabs:Chrome@77.0:macOS 10.14" test.js (Chrome 77.0.3865.75 / macOS 10.14.0)</summary>

```
{
  alias: 'saucelabs:Chrome@77.0:macOS 10.14',
  headless: false,
  name: 'Chrome',
  version: '77.0.3865.75',
  platform: 'desktop',
  os: { name: 'macOS', version: '10.14.0' },
  engine: { name: 'Blink' },
  prettyUserAgent: 'Chrome 77.0.3865.75 / macOS 10.14.0',
  userAgent: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.75 Safari/537.36'
}
```

</details>

<details>

<summary>testcafe "saucelabs:Samsung Galaxy S9 Plus WQHD GoogleAPI Emulator@8.1" test.js (Chrome 67.0.3396.87 / Android 8.1.0)</summary>

```
{
  alias: 'saucelabs:Samsung Galaxy S9 Plus WQHD GoogleAPI Emulator@8.1',
  headless: false,
  name: 'Chrome',
  version: '67.0.3396.87',
  platform: 'mobile',
  os: { name: 'Android', version: '8.1.0' },
  engine: { name: 'Blink' },
  prettyUserAgent: 'Chrome 67.0.3396.87 / Android 8.1.0',
  userAgent: 'Mozilla/5.0 (Linux; Android 8.1.0; Android SDK built for x86 Build/OSM1.180201.026) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/67.0.3396.87 Mobile Safari/537.36'
}

```

</details>

### CODE REVIEW

File: src/client/driver/driver.js [hunk 1]
```diff
@@ -614,6 +618,9 @@ export default class Driver {
         else if (command.type === COMMAND_TYPE.getBrowserConsoleMessages)
             this._onGetBrowserConsoleMessagesCommand(command);
 
+        else if (command.type === COMMAND_TYPE.getBrowserInfo)
```
Comment (inikulin) [inline src/client/driver/driver.js]: Just curious why you decided to implement it as a client command? As far as I understand all the required info can be extracted from the user agent string. Can't this information be stored and provided by `hammerhead` as user agent string is available in the `User-Agent` header during the page request?

Comment (Farfurix) [reply src/client/driver/driver.js]: Hello, @inikulin.

You are right, browser information will be presented as a `testController` property, based on the `user-agent` string. With this "draft/WIP" PR, I just wanted to see CI results. I will update it as soon as possible.

File: src/utils/parse-user-agent.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
+import Bowser from 'bowser';
+
+const DEFAULT_BROWSER_NAME    = 'Other';
+const DEFAULT_BROWSER_VERSION = '0.0';
+// NOTE: we need to pass the ' ' user agent to get a "default" parsed object (GH-481)
+const EMPTY_BOWSER_USER_AGENT = ' ';
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export default function parseUserAgent (userAgent: string, browserAlias?: string): any {
+    let parsedUserAgent;
```
Comment (miherlosev) [inline src/utils/parse-user-agent.ts]: The `browserAlias` parameter is necessary only to return the `.headless` property from the `parseUserAgent` function.
I propose to split this function into 2 two parts:
* `parseUserAgent(userAgent: string)` - parse user agent string and returns the parsed part.
* `isHeadless(alias: string)` - calculated that browser is running in the headless mode or not.

and call both these function in `BrowserConnection.establish` method.

@AndreyBelym What do think about it?

File: test/server/data/execute-async-expression/runtime-error [hunk 3]
```diff
@@ -8,7 +8,7 @@ undefined was passed.
 
 const s = Selector();
 
-Browser: Chrome 15.0.874 / Mac OS X 10.8.1
+Browser: Chrome 77.0.3865.120 / macOS 10.15
```
Comment (miherlosev) [inline test/server/data/execute-async-expression/runtime-error]: Can we revert changes that updates the Chrome up to the latest version (`77`)?

File: src/api/request-hooks/request-logger.js [hunk 4]
```diff
@@ -38,7 +38,7 @@ class RequestLoggerImplementation extends RequestHook {
     }
 
     async onRequest (event) {
-        const userAgent = parseUserAgent(event._requestInfo.userAgent).toString();
+        const userAgent = parseUserAgent(event._requestInfo.userAgent).userAgent;
```
Comment (miherlosev) [inline src/api/request-hooks/request-logger.js]: Let's rename the `userAgent` variable to make the code more clear.
```js
const shortUserAgent = ....
```

File: test/functional/fixtures/api/es-next/browser-info/testcafe-fixtures/browser-info-test.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+    expectedValues.forEach(function (expectedValue) {
+        if (loweredValue === expectedValue)
+            return false;
+
+        return true;
+    });
+}
+
+test
```
Comment (miherlosev) [inline test/functional/fixtures/api/es-next/browser-info/testcafe-fixtures/browser-info-test.js]: At present, you've checked that current browser information value is one of expected values.
I prefer an another testing plan: 
* add server tests for all cases: empty user agent, valid user agent. check `headless` calculation.
* run an empty page only in chrome, retrieve the user agent, parse it with bowser and compare with `t.bowser`.

File: test/functional/fixtures/api/es-next/eval/test.js [hunk 6]
```diff
@@ -1,10 +1,10 @@
 const expect         = require('chai').expect;
-const parseUserAgent = require('useragent').parse;
+const parseUserAgent = require('../../../../../../lib/utils/parse-user-agent');
 const config         = require('../../../../config');
 
 describe('[API] t.eval', function () {
     it('Should execute an anonymous client function', function () {
-        function assertUA (errs, alias, expected) {
+        function assertUA (errs, alias) {
```
Comment (AndreyBelym) [inline test/functional/fixtures/api/es-next/eval/test.js]: Looks like a duplicate of the `assertUA` from

https://github.com/DevExpress/testcafe/pull/4349/files#diff-bbc20d0b3be72d8d04b2a53d69e0dcebR7

Can we extract it to a module?

File: src/utils/parse-user-agent.ts (hunk 2 above)
Comment (AndreyBelym) [reply src/utils/parse-user-agent.ts]: Agreed. Ideally we should use `provider.isHeadlessBrowser()` method to determine if the current browser is headless:

https://github.com/DevExpress/testcafe/blob/master/src/browser/provider/index.ts#L235

`BrowserConnection` can retrieve it immediately after a browser has opened, and `t.browser` can retrieve it via `testRun.browserConnection.headless`.

File: src/api/request-hooks/request-logger.js (hunk 4 above)
Comment (Farfurix) [reply src/api/request-hooks/request-logger.js]: I removed this redundant variable:
```ts
{
    ...
    userAgent: parseUserAgent(event._requestInfo.userAgent).userAgent,
    ...
}
```

Comment (miherlosev) [reply src/api/request-hooks/request-logger.js]: Ok

**Comment by Farfurix** [comment]
@AndreyBelym, OK

**Comment by Farfurix** [comment]
@testcafe-build-bot retest

File: src/api/test-controller/index.js [hunk 7]
```diff
@@ -146,6 +146,14 @@ export default class TestController {
         return this.testRun.fixtureCtx;
     }
 
+    _browser$getter () {
+        return assign({}, this.testRun.browserConnection.browserInfo.parsedUserAgent,
+            {
+                alias:    this.testRun.browserConnection.browserInfo.alias,
```
Comment (miherlosev) [inline src/api/test-controller/index.js]: `Alias` is an internal term. For which user cases does it necessary?

File: test/functional/fixtures/api/es-next/assertions/test.js [hunk 8]
```diff
@@ -4,7 +4,8 @@ describe('[API] Assertions', function () {
     it('Should perform .eql() assertion', function () {
         return runTests('./testcafe-fixtures/assertions-test.js', '.eql() assertion', {
             shouldFail: true,
-            only:       'chrome'
+            only:       'chrome',
```
Comment (miherlosev) [inline test/functional/fixtures/api/es-next/assertions/test.js]: I guess we should not change the current behavior.
For you case, check the expected value using [config](https://github.com/DevExpress/testcafe/pull/4349/files#diff-47530cfffc057e1dfc7fb899799c1694R3).

**Comment by Farfurix** [comment]
@testcafe-build-bot retest

**Comment by Farfurix** [comment]
@testcafe-build-bot retest

File: src/utils/parse-user-agent.ts [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    name: string;
+    version: string;
+    platform: string;
+    os: ParsedComponent;
+    engine: ParsedComponent;
+    prettyUserAgent: string;
+    userAgent: string;
+}
+
+export default function parseUserAgent (userAgent: string = ''): ParsedUserAgent {
```
Comment (miherlosev) [inline src/utils/parse-user-agent.ts]: Make this function more structural:
```js
export default function parseUserAgent (userAgent: string = ''): ParsedUserAgent { 
   if (!userAgent)
        parsedUserAgent = createEmptyParsedUserAgent();
   else
      parsedUserAgent = Bowser.parse(userAgent);

  const os: ParsedComponent     = calculateOperationSystem ();
  const engine: ParsedComponent = calculateEngine ();
  const prettyUserAgent         = calculatePrettyUserAgent ();
  
  return {
    ....
  }
}
```

File: ts-defs-src/test-api/test-controller.d.ts [hunk 10]
```diff
@@ -28,19 +28,58 @@ interface BrowserConsoleMessages {
...
+     */
+    version: string;
+    /**
+     * The platform type.
+     */
+    platform: string;
+    /**
+     * Specifies if the browser is in headless mode.
+     */
+    headless: string;
```
Comment (AndreyBelym) [inline ts-defs-src/test-api/test-controller.d.ts]: Discussed in private.

File: ts-defs-src/test-api/test-controller.d.ts [hunk 11]
```diff
@@ -28,19 +28,58 @@ interface BrowserConsoleMessages {
...
+    /**
+     * The name and version of the operating system.
+     */
+    os: { name?: string; version?: string };
```
Comment (AndreyBelym) [inline ts-defs-src/test-api/test-controller.d.ts]: Discussed in private.

File: ts-defs-src/test-api/test-controller.d.ts [hunk 12]
```diff
@@ -28,19 +28,58 @@ interface BrowserConsoleMessages {
...
+    /**
+     * The name and version of the browser engine.
+     */
+    engine: { name?: string; version?: string };
```
Comment (AndreyBelym) [inline ts-defs-src/test-api/test-controller.d.ts]: Discussed in private.

**Comment by Farfurix** [comment]
@VasilyStrelyaev https://github.com/DevExpress/testcafe/pull/4349/files#diff-3f3d3966c089c8767e453bd78a2a7b77R95:
`Returns an object that contains browser information.`

**Comment by Farfurix** [comment]
@testcafe-build-bot retest
